src/source.rs

```diff
@@ -33,85 +27,62 @@ pub async fn load(
 ) -> Result<SourceManager, Error> {
     let SourceConfiguration { root, paths, includes, excludes, extensions } = configuration;
 
-    let mut starting_paths = Vec::new();
-
-    if paths.is_empty() {
-        starting_paths.push((root.clone(), true));
-    } else {
-        for source in paths {
-            starting_paths.push((source.clone(), true));
-        }
-    }
-
-    for include in includes {
-        starting_paths.push((include.clone(), false));
-    }
-
-    if paths.is_empty() && includes.is_empty() {
-        starting_paths.push((root.clone(), true));
-    }
-
-    let excludes_set: HashSet<Exclusion> = excludes
-        .iter()
-        .map(|exclude| {
-            // if it contains a wildcard, treat it as a pattern
-            if exclude.contains('*') {
-                Exclusion::Pattern(exclude.clone())
-            } else {
-                let path = Path::new(exclude);
-
-                if path.is_absolute() {
-                    Exclusion::Path(path.to_path_buf())
-                } else {
-                    Exclusion::Path(root.join(path))
+    let manager = SourceManager::new(interner.clone());
+    let extensions: HashSet<&String> = extensions.iter().collect();
+    let has_paths = !paths.is_empty();
+    let has_includes = !includes.is_empty();
+    let has_excludes = !excludes.is_empty();
+
+    let entries = jwalk::WalkDir::new(root.clone()).process_read_dir(|_, _, _, children| {
+        children.iter_mut().for_each(|dir_entry_result| {
+            if let Ok(dir_entry) = dir_entry_result {
+                if dir_entry.path().starts_with(".") || dir_entry.file_name.eq_ignore_ascii_case("node_modules") {
+                    dir_entry.read_children_path = None;
                 }
             }
-        })
-        .collect();
+        });
+    });
 
-    let extensions: HashSet<&String> = extensions.iter().collect();
+    for entry in entries {
+        if let Err(_) = entry {
+            continue;
+        }
```

Comment:
we should return the error here instead of ignoring it, if the directory is not readable for example, it should be excluded.

Reply:
What do you suggest?